### ISSUE TITLE
Release builds break D3D12 ray tracing samples.

### ISSUE THREAD

**Issue by crud89** [crud89/LiteFX#136]
**Describe the Bug**

When building in release profile, the D3D12 ray-tracing samples BLAS appears to be broken. This might be caused by compaction and only affects the DirectX 12 backend.

**Comment by Tractorou24** [comment]
It looks like an MSVC error since clang compiles without issue:  
`error C2146: syntax error: missing '>' before identifier '__A0'`  
`error C2511: 'const LiteFX::Rendering::IImage &LiteFX::Rendering::Backends::VulkanFrameBuffer::operator [](unknown-type) const': overloaded member function not found in 'LiteFX::Rendering::Backends::VulkanFrameBuffer'`

__A0 is a reserved identifier, certainly used in compilation. It may be fixed in a later update. If we find a small repro, it can also be reported.

**Comment by crud89** [comment]
Maybe I should have added a little bit more context to this issue, but I guess those observations are different from what I was experiencing. How did you get to those compiler errors?

The problem I am facing with this issue is that ray tracing under the DirectX backend breaks in different ways, but only in release builds. When I first encountered this problem, the BLAS geometry appeared broken, so I created this issue with the intend to take a look at it later. A few days ago, I tried to debug the RT samples by comparing recorded commands between release and debug builds using NSight and PIX. Weirdly enough, the RT sample started working again (without any changes), but in the ray queries sample, no geometry is rendered. The recorded commands in NSight are the same between debug and release builds and I am able to view the whole TLAS/BLAS structure. Debugging the pixel shader in PIX, however, always results in a miss detected on first chance (i.e. no hit detected at all).

This is where I left it, but with the clang support you provided in the PR, I can see that release and debug builds both work. I am still unsure why this happens (again, this only happens with the DirectX backend), but I will try to run this through an MSVC release build with your changes to see if it gets fixed. If not, my current hope is to catch this when writing RT tests for #129.

**Comment by Tractorou24** [comment]
I had those errors Compiling with MSVC in release mode...
Tested it with the latest 17.11 VS preview version.

**Comment by crud89** [comment]
I've looked into the issue, but it appears unrelated. It is fixed when removing the covariance from the return value of the `operator[]` and `image` functions of the `FrameBuffer` implementations:

```cxx
//inline const IVulkanImage& operator[](StringView renderTargetName) const override {
inline const IImage& operator[](StringView renderTargetName) const override {
    return this->resolveImage(hash(renderTargetName));
}

//inline const IVulkanImage& image(StringView renderTargetName) const override {
inline const IImage& image(StringView renderTargetName) const override {
    return this->resolveImage(hash(renderTargetName));
}
```

This indeed appears like a regression with the compiler. I will see if I can create a MRE from it and report it. As for the engine, I will try to move this to the base class and see if it resolves the issue.

Nevertheless, ray queries are still broken. I've managed to create two identical PIX captures, one from the release and the other one from the debug builds. Both use the same shaders (debug version, as they retain the PDBs). I've spent quite some time investigating differences, but couldn't find any. I've tested the samples on a 2080 Ti, where RT works, but RQs always return `COMMITTED_NOTHING`. On my 4080, however, both ray tracing samples show invalid geometry. I suspect this is somehow related to how acceleration structures are built in the D3D backend.

Interestingly *RelWithDebInfo* also works, so just MSVC release is broken.

**Comment by crud89** [comment]
To my pleasent surprise, this indeed appears to have been a codegen issue. I will have to check on another system yet, but it seems that this got fixed with VS 17.12. At least on one of my dev machine the samples now all work with the current main branch.

However, this was a different issue to the syntax error produced in VS 17.11 before, as this got also fixed and backported to 17.11.5, but did not fix the ray query samples.

In any case this still seems like a mystery to me. On the one hand every other compiler (GCC, clang, clang-cl) produced working release builds. Even MSVC generated a working build with the `RelWithDebInfo` profile. This strongly points towards the compiler. However, I do have a PIX capture that shows that the acceleration structures are built properly - or at least I couldn't identify any issues with it - and it's just the query part that always misses (i.e., always hits `COMMITTED_NOTHING`). That strongly points towards the code. My current hypothesis is still that we are invoking some sort of UB somewhere.

I will leave the PIX captures below. They contain the first 3 frames of the ray queries sample, both using identical shader builds (including debug information). If anybody is willing to pick this up again, be my guest! A hint to whats different between the captures would help a ton.

- [Debug build (working)](https://github.com/user-attachments/files/17729900/rq_debug.zip)
- [Release build (faulty)](https://github.com/user-attachments/files/17729895/rq_release.zip)
